_posts/2024-06-19-optimizing-opensearch-with-fp16-quantization.md

```diff
@@ -0,0 +1,287 @@
+|---	|---	|---	|
+|OpenSearch version	|2.13	|2.13	|
+|engine	|faiss	|faiss	|
+|Dimension of vector	|768	|768	|
+|ingest vectors	|100M	|100M	|
+|test vectors	|1k	|1k	|
+|primary shards	|36	|36	|
+|replica shards	|0	|0	|
+|data nodes	|4	|8	|
+|data node instance type	|r5.4xlarge	|r5.4xlarge	|
+|master nodes	|3	|3	|
+|master node instance type	|c5.xlarge	|c5.xlarge	|
+|indexing clients	|9	|9	|
+|query clients	|1	|1	|
+|forcemerge segments	|1	|1	|
+|client instance	|r5.16xlarge	|r5.16xlarge	|
+
+Config ID	|Optimization Strategy	|m	|ef_construction	|ef_search	|
+|---	|---	|---	|---	|---	|
+|hnsw1	|Default configuration	|16	|100	|100	|
+|hnsw2	|Balance between latency, memory, and recall	|16	|128	|128	|
+|hnsw3	|Optimize for recall	|16	|256	|256	|
+
+Faiss HNSW SQfp16 requires 4 data nodes, half the number needed for Faiss HNSW (8). This demonstrates that SQfp16 reduces memory requirements by 50%. 
+For more information about estimating required memory and number of data nodes, see the [Appendix](#appendix-memory-and-data-node-requirement-estimation).
+
+#### Recall and memory results
+
+|Experiment ID	|hnsw-recall@1000	|hnsw-sqfp16-recall@1000	|hnsw memory usage (gb)	|hnsw-sqfp16 memory usage (gb)	|% reduction in memory	|
+|---	|---	|---	|---	|---	|---	|
+|hnsw 1	|0.94	|0.94	|300.28	|157.23	|47.64	|
+|hnsw 2	|0.96	|0.96	|300.28	|157.23	|47.64	|
+|hnsw 3	|0.98	|0.98	|300.28	|157.23	|47.64	|
+
+#### Indexing and query results
+
+|Experiment ID	|hnsw mean throughput (docs/sec)	|hnsw-sqfp16 mean throughput (docs/sec)	|hnsw p90 (ms)	|hnsw-sqfp16 p90 (ms)	|hnsw p99 (ms)	|hnsw-sqfp16 p99 (ms)	|
+|---	|---	|---	|---	|---	|---	|---	|
+|hnsw 1	|7544	|7657	|14.02	|16.99	|19.18	|20.83	|
+|hnsw 2	|7063	|7219	|14.21	|17.44	|18.86	|21.80	|
+|hnsw 3	|6004	|5848	|16.14	|20.85	|17.65	|24.73	|
+
+#### Analysis
+
+* For k=1000, the recall is identical for both Faiss HNSW and Faiss HNSW with SQfp16.
+* Faiss HNSW with SQfp16 requires approximately half the memory Faiss HNSW required (as measured by the required number of data nodes). Based on the [k-NN stats API metrics](https://opensearch.org/docs/latest/search-plugins/knn/api/#stats), the memory usage has been reduced by 47.64% using SQfp16.
+* In most instances, SQfp16 demonstrated better indexing throughput compared to `fp32` vectors.
+
+## Conclusion
+
+Faiss FP16 scalar quantization is a powerful technique that provides significant memory savings while maintaining high recall performance similar to full-precision vectors. By converting vectors to a 16-bit floating-point representation, it can reduce memory requirements by up to 50%. When combined with SIMD (Single Instruction Multiple Data) optimization, FP16 scalar quantization also enhances indexing throughput and reduces search latencies, leading to better overall performance. This method strikes an excellent balance between memory efficiency and accuracy, making it a valuable tool for large-scale similarity search applications.
```

Comment:
```suggestion
Faiss FP16 scalar quantization is a powerful technique that provides significant memory savings while maintaining high recall performance similar to full-precision vectors. By converting vectors to a 16-bit floating-point representation, it can reduce memory requirements by up to 50%. When combined with SIMD optimization, FP16 scalar quantization also enhances indexing throughput and reduces search latencies, leading to better overall performance. This method strikes an excellent balance between memory efficiency and accuracy, making it a valuable tool for large-scale similarity search applications.
```